File: app/src/main/java/com/example/tesonet/activities/ServerListActivity.java
package com.example.tesonet.activities;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.tesonet.R;
import com.example.tesonet.adapter.ServersAdapter;
import com.example.tesonet.database.models.Server;
import com.example.tesonet.viewmodels.ServerListViewModel;

import java.util.List;

public class ServerListActivity extends AppCompatActivity {
    ServerListViewModel serverListViewModel;
    ServersAdapter serversAdapter;
    ConstraintLayout serversListConstraintLayout;
    ConstraintLayout loadingConstraintLayout;
    TextView loaderTextView;
    String tokenString;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_server_list);

        Intent intent = getIntent();
        tokenString = intent.getStringExtra("token");

        init();
    }

    public void logout(View view) {
        destroyActivity();
    }

    @Override
    public void onBackPressed() {
        //Paliekam taip, paspaudus atgal nieko neivyks.
        //Galima iskviesti alert dialog ir pasiulyti atsijungti.
    }

    private void init() {
        //RecyclerView
        final RecyclerView recyclerView = findViewById(R.id.recycleView);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setHasFixedSize(true);

        //ConstraintLayouts
        serversListConstraintLayout = findViewById(R.id.serverListConstraintLayout);
        loadingConstraintLayout = findViewById(R.id.loadingConstraintLayout);

        loadingScreen();

        //TextViews
        loaderTextView = findViewById(R.id.loaderTextView);

        //ViewModel
        serverListViewModel = ViewModelProviders.of(this).get(ServerListViewModel.class);
        serverListViewModel.setToken(tokenString);
        serverListViewModel.getServerList().observe(this, new Observer<List<Server>>() {
            @Override
            public void onChanged(@Nullable List<Server> servers) {
                serversAdapter = new ServersAdapter(ServerListActivity.this, servers);
                recyclerView.setAdapter(serversAdapter);
                serversListConstraintLayout.setVisibility(View.VISIBLE);
                loadingConstraintLayout.setVisibility(View.INVISIBLE);
            }
        });

        serverListViewModel.getError().observe(this, new Observer<String>() {
            @Override
            public void onChanged(String s) {
                if(s.equals("error")){
                    destroyActivity();
                }
            }
        });
    }

    public void loadingScreen() {
        //Loading screen init
        loadingConstraintLayout.setVisibility(View.VISIBLE);
        serversListConstraintLayout.setVisibility(View.INVISIBLE);
    }

    public void destroyActivity() {
        serverListViewModel.deleteServers();
        finish();
    }
}

File: app/src/main/java/gj/tesonet/ui/servers/ServerListActivity.kt
package gj.tesonet.ui.servers

import android.content.Context
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.observe
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import gj.tesonet.R
import gj.tesonet.ui.login.LoginActivity
import gj.tesonet.ui.show
import kotlinx.android.synthetic.main.activity_servers_list.*

class ServerListActivity : AppCompatActivity() {

    private val adapter: ServerListAdapter by lazy {
        ServerListAdapter()
    }

    private val model: ServerListViewModel by lazy {
        ViewModelProvider(this).get(ServerListViewModel::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_servers_list)

        list.layoutManager = LinearLayoutManager(this)
        list.adapter = adapter
        list.addItemDecoration(
            DividerItemDecoration(this, DividerItemDecoration.VERTICAL).apply {
                getDrawable(R.drawable.divider)?.let {
                    setDrawable(it)
                }
            }
        )

        model.servers.observe(this) { servers ->
            adapter.list = servers

            if (servers != null) {
                background.visibility = View.GONE
                list.visibility = View.VISIBLE
                header.visibility = View.VISIBLE
            }
        }

        model.message.observe(this) { reference ->
            reference.getAndSet(null)?.let { message ->
                show(message)
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_servers, menu)

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item?.itemId == R.id.logout) {
            model.logout()

            LoginActivity.start(this)
            finish()
        }

        return super.onOptionsItemSelected(item)
    }

    companion object {

        fun start(context: Context) {
            val intent = Intent(context, ServerListActivity::class.java)

            context.startActivity(intent)
        }
    }
}
